### PR TITLE
Auto-Calibrating Maintainer

### PR DESCRIPTION
**Pull request by Gabriel-J-Young**
added calibrate, maintainer, and manual modes along with tests for the calibrate command.

### CODE REVIEW

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java [hunk 1]
```diff
@@ -3,18 +3,33 @@
...
+import xbot.common.properties.XPropertyManager;
+import competition.operator_interface.OperatorInterface;
 
 public class ElevatorMaintainerCommand extends BaseCommand {
 
+    OperatorInterface oi;
     ElevatorSubsystem elevator;
     PIDManager pid;
+    double startTime;
+    double attemptTime;
```
Comment (JohnGilb) [inline Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java]: Rename this to `giveUpCalibratingTime`

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java [hunk 2]
```diff
@@ -3,18 +3,33 @@
...
+    final DoubleProperty calibrationAttemptTime;
 
     @Inject
-    public ElevatorMaintainerCommand(ElevatorSubsystem elevator, PIDFactory pf) {
+    public ElevatorMaintainerCommand(
+            ElevatorSubsystem elevator, 
+            PIDFactory pf, 
+            XPropertyManager propMan, 
+            OperatorInterface oi) {
+        calibrationAttemptTime = propMan.createPersistentProperty("Calibration attempt time (ms)", 4000);
```
Comment (JohnGilb) [inline Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java]: `Elevator Calibration Attempt Time (ms)`

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java [hunk 3]
```diff
@@ -3,18 +3,33 @@
...
+import xbot.common.properties.DoubleProperty;
+import xbot.common.properties.XPropertyManager;
+import competition.operator_interface.OperatorInterface;
 
 public class ElevatorMaintainerCommand extends BaseCommand {
 
+    OperatorInterface oi;
     ElevatorSubsystem elevator;
     PIDManager pid;
+    double startTime;
```
Comment (JohnGilb) [inline Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java]: Rename to `initializationTime`

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java [hunk 4]
```diff
@@ -24,16 +39,36 @@ public void initialize() {
         log.info("Initializing");
         if (!elevator.isCalibrated()) {
             log.warn("ELEVATOR UNCALIBRATED - THIS COMMAND WILL NOT DO ANYTHING!");
+            startTime = Timer.getFPGATimestamp();
+            attemptTime = startTime + calibrationAttemptTime.get();
```
Comment (JohnGilb) [inline Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java]: add a line with
`log.info("Attempting calibration until " + attemptTime)`

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java [hunk 5]
```diff
@@ -24,16 +39,36 @@ public void initialize() {
...
         }
     }
 
     @Override
     public void execute() {
+        boolean maintain = false;
```
Comment (JohnGilb) [inline Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java:54]: Let's replace the three booleans with an enum - we'll talk more about it on Tuesday, but I suggest you do some reading on enums in the meantime.

**Review by JohnGilb — CHANGES_REQUESTED** [review]

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java [hunk 6]
```diff
@@ -24,16 +39,36 @@ public void initialize() {
...
+            manual = true;
+        }
+        
+        
+        
+        if (maintain) {
+            elevator.setPower(pid.calculate(elevator.getTargetHeight(), elevator.getCurrentHeightInInches()));
+        }
+        else if (tryToCalibrate) {
+            elevator.setPower(-1);
```
Comment (WasabiFan) [inline Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java]: This should be a property

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java (hunk 1 above)
Comment (Gabriel-J-Young) [reply Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java]: done

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java (hunk 2 above)
Comment (Gabriel-J-Young) [reply Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java]: renamed this to "elevatorCalibrationAttemptTimeMS" I'm not sure how to properly use the backslash to deal with parenthesizes.

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java (hunk 3 above)
Comment (Gabriel-J-Young) [reply Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java]: I removed this variable and just called the timer in initialize to get the giveUpCalibrationTime.

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java (hunk 4 above)
Comment (Gabriel-J-Young) [reply Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java]: added a line with log.info("Attempting calibration from" + Timer.getFPGATimestamp() + " until " + giveUpCalibrationTime")

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java (hunk 6 above)
Comment (Gabriel-J-Young) [reply Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java]: what specifically?

Comment (JohnGilb) [reply Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java]: @WasabiFan I don't think it's currently relevant - the Elevator subsystem already has power restrictions if it's uncalibrated, which are controlled by the DoubleProperty `calibrationPower`

Comment (JohnGilb) [reply Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java]: To clean this up though, we could replace this line with `elevator.lower()` so it's always controlled by some property.

Comment (WasabiFan) [reply Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java]: I don't think that is the right solution. Relying on an assumed internal behavior of the subsystem for this command's functionality isn't safe and confuses a reader. If the subsystem universally limits power while it's uncalibrated, that should be called `Uncalibrated max power` and this command should have a separate property for the speed it uses to calibrate.

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java (hunk 5 above)
Comment (Gabriel-J-Young) [reply Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java:54]: done

File: Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java (hunk 6 above)
Comment (Gabriel-J-Young) [reply Competition2018/src/main/java/competition/subsystems/elevator/commands/ElevatorMaintainerCommand.java]: changed "elevator.setPower(-1)" to "elevator.lower()"

**Review by JohnGilb — APPROVED** [review]

**Pull request merged**